mypy error with Tagged Unions

👋 I'm hitting a mypy error: Here's my simple test:

```python
# msgspec_test.py
from msgspec import Struct
from msgspec.json import encode, decode
from typing import Union

class Automation(Struct):
    name: str
    args: dict = {}

class TemperatureAutomation(Automation, tag="temperature"):
    ...

class DosingAutomation(Automation, tag="dosing"):
    ...

class LEDAutomation(Automation, tag="led"):
    ...


value = encode(LEDAutomation(name="thing", args={}))
decode(value, type=Union[TemperatureAutomation, DosingAutomation, LEDAutomation])
```

Running `mypy msgspect_test.py` gives:

```
msgspec_test.py:20: error: No overload variant of "decode" matches argument types "bytes", "object"
msgspec_test.py:20: note: Possible overload variants:
msgspec_test.py:20: note:     def decode(buf: bytes, *, dec_hook: Optional[Callable[[Type[Any], Any], Any]] = ...) -> Any
msgspec_test.py:20: note:     def [T] decode(buf: bytes, *, type: Type[T] = ..., dec_hook: Optional[Callable[[Type[Any], Any], Any]] = ...) -> T
Found 1 error in 1 file (checked 1 source file)
```

It works fine with pyright though.

### Versions

mypy==0.971
msgspec==0.8.0
pyright==1.1.268

--- a/msgspec/json.pyi
+++ b/msgspec/json.pyi
@@ -33,6 +33,7 @@ class Encoder:
 class Decoder(Generic[T]):
     type: Type[T]
     dec_hook: dec_hook_sig
+
     @overload
     def __init__(
         self: Decoder[Any],
@@ -43,6 +44,13 @@ class Decoder(Generic[T]):
     def __init__(
         self: Decoder[T],
         type: Type[T] = ...,
+        *,
+        dec_hook: dec_hook_sig = None,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self: Decoder[Any],
+        type: Any = ...,
         *,
         dec_hook: dec_hook_sig = None,
     ) -> None: ...
@@ -61,6 +69,13 @@ def decode(
     type: Type[T] = ...,
     dec_hook: dec_hook_sig = None,
 ) -> T: ...
+@overload
+def decode(
+    buf: bytes,
+    *,
+    type: Any = ...,
+    dec_hook: dec_hook_sig = None,
+) -> Any: ...
 def encode(obj: Any, *, enc_hook: enc_hook_sig = None) -> bytes: ...
 def schema(type: Any) -> Dict[str, Any]: ...
 def schema_components(

--- a/msgspec/msgpack.pyi
+++ b/msgspec/msgpack.pyi
@@ -32,6 +32,14 @@ class Decoder(Generic[T]):
         dec_hook: dec_hook_sig = None,
         ext_hook: ext_hook_sig = None,
     ) -> None: ...
+    @overload
+    def __init__(
+        self: Decoder[Any],
+        type: Any = ...,
+        *,
+        dec_hook: dec_hook_sig = None,
+        ext_hook: ext_hook_sig = None,
+    ) -> None: ...
     def decode(self, data: bytes) -> T: ...
 
 class Encoder:
@@ -63,4 +71,12 @@ def decode(
     dec_hook: dec_hook_sig = None,
     ext_hook: ext_hook_sig = None,
 ) -> T: ...
+@overload
+def decode(
+    buf: bytes,
+    *,
+    type: Any = ...,
+    dec_hook: dec_hook_sig = None,
+    ext_hook: ext_hook_sig = None,
+) -> Any: ...
 def encode(obj: Any, *, enc_hook: enc_hook_sig = None) -> bytes: ...

--- a/tests/basic_typing_examples.py
+++ b/tests/basic_typing_examples.py
@@ -1,7 +1,9 @@
 # fmt: off
+from __future__ import annotations
+
 import datetime
 import pickle
-from typing import List, Any, Type
+from typing import List, Any, Type, Union
 import msgspec
 
 def check___version__() -> None:
@@ -391,6 +393,14 @@ def check_msgpack_Decoder_decode_typed() -> None:
     reveal_type(o)  # assert ("List" in typ or "list" in typ) and "int" in typ
 
 
+def check_msgpack_Decoder_decode_union() -> None:
+    # Pyright doesn't require the annotation, but mypy does until TypeForm
+    # is supported. This is mostly checking that no error happens here.
+    dec: msgspec.msgpack.Decoder[Union[int, str]] = msgspec.msgpack.Decoder(Union[int, str])
+    o = dec.decode(b'')
+    reveal_type(o)  # assert ("int" in typ and "str" in typ)
+
+
 def check_msgpack_Decoder_decode_type_comment() -> None:
     dec = msgspec.msgpack.Decoder()  # type: msgspec.msgpack.Decoder[List[int]]
     b = msgspec.msgpack.encode([1, 2, 3])
@@ -412,6 +422,11 @@ def check_msgpack_decode_typed() -> None:
     o = msgspec.msgpack.decode(b, type=List[int])
 
     reveal_type(o)  # assert ("List" in typ or "list" in typ) and "int" in typ
+
+
+def check_msgpack_decode_typed_union() -> None:
+    o: Union[int, str] = msgspec.msgpack.decode(b"", type=Union[int, str])
+    reveal_type(o)  # assert "int" in typ and "str" in typ
 
 
 def check_msgpack_encode_enc_hook() -> None:
@@ -495,6 +510,12 @@ def check_json_Decoder_decode_type_comment() -> None:
     reveal_type(o)  # assert ("List" in typ or "list" in typ) and "int" in typ
 
 
+def check_json_Decoder_decode_union() -> None:
+    dec: msgspec.json.Decoder[Union[int, str]] = msgspec.json.Decoder(Union[int, str])
+    o = dec.decode(b'')
+    reveal_type(o)  # assert ("int" in typ and "str" in typ)
+
+
 def check_json_decode_any() -> None:
     b = msgspec.json.encode([1, 2, 3])
     o = msgspec.json.decode(b)
@@ -507,6 +528,11 @@ def check_json_decode_typed() -> None:
     o = msgspec.json.decode(b, type=List[int])
 
     reveal_type(o)  # assert ("List" in typ or "list" in typ) and "int" in typ
+
+
+def check_json_decode_typed_union() -> None:
+    o: Union[int, str] = msgspec.json.decode(b"", type=Union[int, str])
+    reveal_type(o)  # assert "int" in typ and "str" in typ
 
 
 def check_json_encode_enc_hook() -> None:

